packages/angular_devkit/core/src/analytics/forwarder.ts

```diff
@@ -0,0 +1,131 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { JsonObject } from '../json';
+import { Analytics, EventOptions, PageviewOptions, ScreenviewOptions, TimingOptions } from './api';
+
+
+export enum AnalyticsReportKind {
+  Event = 'event',
+  Screenview = 'screenview',
+  Pageview = 'pageview',
+  Timing = 'timing',
+}
+
+export interface AnalyticsReportBase extends JsonObject {
+  kind: AnalyticsReportKind;
+}
+
+export interface AnalyticsReportEvent extends AnalyticsReportBase {
+  kind: AnalyticsReportKind.Event;
+  options: JsonObject & EventOptions;
+  c: string;
+  a: string;
+}
+export interface AnalyticsReportScreenview extends AnalyticsReportBase {
+  kind: AnalyticsReportKind.Screenview;
+  options: JsonObject & ScreenviewOptions;
+  s: string;
+  a: string;
+}
+export interface AnalyticsReportPageview extends AnalyticsReportBase {
+  kind: AnalyticsReportKind.Pageview;
+  options: JsonObject & PageviewOptions;
+  p: string;
+}
+export interface AnalyticsReportTiming extends AnalyticsReportBase {
+  kind: AnalyticsReportKind.Timing;
+  options: JsonObject & TimingOptions;
+  c: string;
+  v: string;
+  t: string | number;
+}
+
+export type AnalyticsReport =
+  AnalyticsReportEvent
+  | AnalyticsReportScreenview
+  | AnalyticsReportPageview
+  | AnalyticsReportTiming
+  ;
+
+/**
+ * A function that can forward analytics along some stream. AnalyticsReport is already a
+ * JsonObject descendant, but we force it here so the user knows it's safe to serialize.
+ */
+export type AnalyticsForwarderFn = (report: JsonObject & AnalyticsReport) => void;
+
+/**
+ * A class that follows the Analytics interface and forward analytic reports (JavaScript objects).
+ * AnalyticsReporter is the counterpart which takes analytic reports and report them to another
+ * Analytics interface.
+ */
+export class AnalyticsForwarder implements Analytics {
+  constructor(protected _fn: AnalyticsForwarderFn) {}
+
+  event(category: string, action: string, options?: EventOptions) {
+    this._fn({
+      kind: AnalyticsReportKind.Event,
+      c: category,
+      a: action,
+      options: { ...options } as JsonObject,
+    });
+  }
+  screenview(screenName: string, appName: string, options?: ScreenviewOptions) {
+    this._fn({
+      kind: AnalyticsReportKind.Screenview,
+      s: screenName,
+      a: appName,
+      options: { ...options } as JsonObject,
+    });
+  }
+  pageview(path: string, options?: PageviewOptions) {
+    this._fn({
+      kind: AnalyticsReportKind.Pageview,
+      p: path,
+      options: { ...options } as JsonObject,
+    });
+  }
+  timing(category: string, variable: string, time: string | number, options?: TimingOptions): void {
+    this._fn({
+      kind: AnalyticsReportKind.Timing,
+      c: category,
+      v: variable,
+      t: time,
+      options: { ...options } as JsonObject,
+    });
+  }
+
+  // We do not support flushing.
+  flush(): never {
+    throw new Error('Flush is not supported.');
```

Comment:
I find this pattern a bit confusing:
- `AnalyticsForwarder` doesn't support flushing and throws
- `LoggingAnalytics` and `NoopAnalytics` just resolve a promise without doing anything
- `MultiAnalytics` flushes all of its backends

There are 3 other `Analytics` implementations for `MultiAnalytics` to use. Two of them do nothing, and one of them throws. 

I think this leaves flushing with `MultiAnalytics` analytics in an odd place since it will always fail if there's at least one `AnalyticsForwarder`, and you can't really know which one failed, or if multiple failed, or how to proceed.

Maybe flushing should instead do nothing when it's not supported?

Reply:
You're right, it should do nothing or flush. Ideally it should actually flush and wait for the other side to be done, but that requires two-way communication and a lot more plumbing than I'm comfortable with. So it will be a do nothing and hope for the best.